### PR TITLE
Add Provider API script for Freesound

### LINKED ISSUE #16: [API Integration - AUDIO] FreeSound.org (original #282)
This issue has been migrated from the [CC Search Catalog repository](https://github.com/creativecommons/cccatalog/issues/282)

```
Author: annatuma
Date: Tue Feb 25 2020
Labels: providers,✨ goal: improvement,🙅 status: discontinued
```
**This ticket is blocked until the CC Catalog can support ingesting audio as a content type. Further ticket work will also be required.**

## Provider API Endpoint / Documentation

https://freesound.org/help/developers/
https://freesound.org/docs/api/
https://freesound.org/apiv2/

For this provider, an API key is required. It is not possible to use CC’s org key, as a community contributor. The preferred way to work with it is to use an environment variable, such as PROVIDERNAME_API_KEY for the key and read it in the script.

## Licenses Provided

Their main information page indicates that all sounds are under CC license, however, we should verify this/restrict by license info, which is possible using the "license" field in
GET /apiv2/sounds/<sound_id>/

## General Recommendations for implementation

- The script should be in the `src/cc_catalog_airflow/dags/provider_api_scripts/` directory.
- The script should have a test suite in the same directory.
- The script must use the `ImageStore` class (Import this from
  `src/cc_catalog_airflow/dags/provider_api_scripts/common/storage/image.py`).
- The script should use the `DelayedRequester` class (Import this from
  `src/cc_catalog_airflow/dags/provider_api_scripts/common/requester.py`).
- The script must not use anything from
  `src/cc_catalog_airflow/dags/provider_api_scripts/modules/etlMods.py`, since
  that module is deprecated.
- If the provider API has can be queried by 'upload date' or something similar,
  the script should take a `--date` parameter when run as a script, giving the
  date for which we should collect images. The form should be `YYYY-MM-DD` (so,
  the script can be run via `python my_favorite_provider.py --date 2018-01-01`).
- The script must provide a main function that takes the same parameters as from
  the CLI. In our example from above, we'd then have a main function
  `my_favorite_provider.main(date)`. The main should do the same thing calling
  from the CLI would do.
- The script *must* conform to [PEP8][pep8]. Please use `pycodestyle` (available via
  `pip install pycodestyle`) to check for compliance.
- The script should use small, testable functions.
- The test suite for the script may break PEP8 rules regarding long lines where
  appropriate (e.g., long strings for testing).

[pep8]: https://www.python.org/dev/peps/pep-0008/

## Examples of other Provider API Scripts

For example Provider API Scripts and accompanying test suites, please see

- `src/cc_catalog_airflow/dags/provider_api_scripts/flickr.py` and
- `src/cc_catalog_airflow/dags/provider_api_scripts/test_flickr.py`, or
- `src/cc_catalog_airflow/dags/provider_api_scripts/wikimedia_commons.py` and
- `src/cc_catalog_airflow/dags/provider_api_scripts/test_wikimedia_commons.py`.

### PR DESCRIPTION
**Pull request by obulat**
Fixes #16

## Provider API Endpoint / Documentation

https://freesound.org/help/developers/
https://freesound.org/docs/api/
https://freesound.org/apiv2/

## Provider description

Freesound aims to create a huge collaborative database of audio snippets, samples, recordings released under Creative Commons licenses that allow their reuse. Its aim is to create an open database of sounds that can also be used for scientific research and be integrated in third party applications.  Freesound was started in 2005 in the Music Technology Group of Universitat Pompeu Fabra, Barcelona, Spain.

## Licenses Provided

There are CC licenses, including CC0, as well as some other open licenses such as [Free Arts License](https://en.wikipedia.org/wiki/Free_Art_License).

## Provider API Technical info

The API provides links to 4 audio files: the main audio that was uploaded by the user, and 4 `preview` files (mp3 and ogg, both in hq and in lq) which are of the same length. We save the main file in `alt_files`. It can be downloaded by an a user who is authenticated at freesound.org. The main file in our database is the `mp3 hq` which is used for playback  
The script uses auth workflow to get the URL for downloadable file.

### CODE REVIEW

**Comment by zackkrida** [comment]
@obulat What do we need here to unblock from draft?

**Comment by obulat** [comment]
I am not sure which URL to use as the main audio URL:
The _main download URL_, usually for a WAV type, requires OAuth2 authentication, which is not so easy to do as a part of an Airflow workflow. Sample URL: https://freesound.org/apiv2/sounds/401082/download/
There also are _mp3 previews_ that seem to open in incognito windows, although the API docs state that you need authorization (token or Oauth2) to get them. Example: https://freesound.org/data/previews/401/401082_7725148-hq.mp3

Here is a sample of what information we can get from the freesound API:
```
{
  "title": "Click_08.wav",
  "creator": "s-cheremisinov",
  "creator_url": "https://freesound.org/people/s-cheremisinov/",
  "foreign_identifier": 401082,
  "foreign_landing_url": "https://freesound.org/people/s-cheremisinov/sounds/401082/",
  "audio_url": "https://freesound.org/people/s-cheremisinov/sounds/401082/",
  "duration": 500,
  "license_info": [
    "cc0",
    "1.0",
    "https://creativecommons.org/publicdomain/zero/1.0/",
    "http://creativecommons.org/publicdomain/zero/1.0/"
  ],
  "meta_data": {
    "description": "Digital design sound of pressing a mouse button.\r\nThis pack of sounds and transitions is made using the software \"Ableton Live\" and it is completely digital, without live recording. Exceptionally sound design. Made in early autumn of 2017. It is ideal for any video and motion design work. I made it as a sign of respect for my friends working with Videohive.",
    "num_downloads": 32,
    "download": "https://freesound.org/apiv2/sounds/401082/download/",
    "previews": {
      "preview-lq-ogg": "https://freesound.org/data/previews/401/401082_7725148-lq.ogg",
      "preview-lq-mp3": "https://freesound.org/data/previews/401/401082_7725148-lq.mp3",
      "preview-hq-ogg": "https://freesound.org/data/previews/401/401082_7725148-hq.ogg",
      "preview-hq-mp3": "https://freesound.org/data/previews/401/401082_7725148-hq.mp3"
    }
  },
  "raw_tags": [
    "button",
    "mouse",
    "click"
  ],
  "audio_set": "s-cheremisinov Free Sound Pack",
  "set_url": "https://freesound.org/apiv2/packs/22598/",
  "alt_files": [
    {
      "url": "https://freesound.org/apiv2/sounds/401082/download/",
      "format": "wav",
      "filesize": 88244,
      "bit_rate": 1379,
      "sample_rate": 44100.0
    },
    {
      "url": "https://freesound.org/data/previews/401/401082_7725148-lq.ogg",
      "format": "ogg"
    },
    {
      "url": "https://freesound.org/data/previews/401/401082_7725148-lq.mp3",
      "format": "mp3"
    },
    {
      "url": "https://freesound.org/data/previews/401/401082_7725148-hq.ogg",
      "format": "ogg"
    },
    {
      "url": "https://freesound.org/data/previews/401/401082_7725148-hq.mp3",
      "format": "mp3"
    }
  ],
  "bit_rate": 1379,
  "sample_rate": 44100
}
```

**Comment by zackkrida** [comment]
Have we confirmed the download link doesn't work with token auth? 

https://freesound.org/docs/api/authentication.html#authentication

Otherwise, we might need to make a dag that simply keeps the oauth token alive at whatever duration they last, which would be annoying but tolerable.

**Comment by obulat** [comment]
> Have we confirmed the download link doesn't work with token auth?

Yes, if you try opening the download link when not signed in, you get redirected to a sign in page.

File: openverse_catalog/dags/provider_api_scripts/freesound.py
Comment (zackkrida) [inline openverse_catalog/dags/provider_api_scripts/freesound.py:203]: Is this structure for `alt_files` documented anywhere? Since it's a JSON column, and this is basically it's type signature, is this enforced at all or could provider scripts set this incorrectly?

**Comment by zackkrida** [comment]
@obulat related discussion on the Oauth2 flow here: https://github.com/WordPress/openverse-catalog/discussions/208

It's would be nice to have one more pair of 👁️s on that!

**Comment by zackkrida** [comment]
@obulat looks like there's some new feedback here and now that you've gotten a reply from Frederic it'd be great to pick this work back up. It's looking great so far 👍

File: openverse_catalog/dags/provider_api_scripts/freesound.py
Comment (obulat) [reply openverse_catalog/dags/provider_api_scripts/freesound.py]: Sorry for such a late reply. I've changed the way files are saved now: the main file (ie the response's `audio_url`, `filesize`, `filetype` are for the `hq` preview in mp3. All other files (including the actual main file in freesound that requires being logged in to download, and other preview files) are saved as altfiles.

**Comment by obulat** [comment]
> @obulat looks like there's some new feedback here and now that you've gotten a reply from Frederic it'd be great to pick this work back up. It's looking great so far 👍

@zackkrida, I've changed the way files are saved now: the main file is the `hq` mp3 preview, all other files are saved as `altfiles`. I've also added the `head` request to get the main file's size. It should probably be removed later to not overload the API, but I think we need to have it for the initial ingestion.

File: openverse_catalog/dags/provider_api_scripts/freesound.py
Comment (sarayourfriend) [reply openverse_catalog/dags/provider_api_scripts/freesound.py]: If the other preview files are saved as altfiles, how will the frontend distinguish between the downloadable one and the other previews?

**Comment by zackkrida** [comment]
Seeing this locally @obulat

CleanShot 2021-11-08 at 16 16 35@2x

File: openverse_catalog/dags/providers/provider_api_scripts/freesound.py
Comment (krysal) [inline openverse_catalog/dags/providers/provider_api_scripts/freesound.py:67]: As @AetherUnbound commented before, it doesn't use a date param so it actually pulls all the data. Better to update to avoid confusion.

**Comment by zackkrida** [comment]
Should we have a more explicit error message when authentication fails? CC @AetherUnbound. Here's the current output:

```
[2021-11-10 20:35:52,020] {freesound.py:69} INFO - Begin: Freesound script
[2021-11-10 20:35:52,745] {requester.py:48} WARNING - Unable to request URL: https://freesound.org/apiv2/search/text/?format=json&token=not_set&query=&page_size=150&fields=id%2Curl%2Cname%2Ctags%2Cdescription%2Ccreated%2Clicense%2Ctype%2Cdownload%2Cfilesize%2Cbitrate%2Cbitdepth%2Cduration%2Csamplerate%2Cpack%2Cusername%2Cnum_downloads%2Cavg_rating%2Cnum_ratings%2Cgeotag%2Cpreviews&page=1&license=Attribution.  Status code: 401
[2021-11-10 20:35:52,745] {requester.py:82} WARNING - Bad response_json:  None
[2021-11-10 20:35:52,745] {requester.py:83} WARNING - Retrying:
_get_response_json(
    https://freesound.org/apiv2/search/text,
    {'format': 'json', 'token': 'not_set', 'query': '', 'page_size': 150, 'fields': 'id,url,name,tags,description,created,license,type,download,filesize,bitrate,bitdepth,duration,samplerate,pack,username,num_downloads,avg_rating,num_ratings,geotag,previews', 'page': '1', 'license': 'Attribution'},
    retries=2)
```

File: openverse_catalog/dags/providers/provider_api_scripts/freesound.py
Comment (krysal) [inline openverse_catalog/dags/providers/provider_api_scripts/freesound.py:169]: Seems like this is still a WIP (?) I believe these fields are derived from `main_audio` here.

**Review by krysal — COMMENTED** [review]
It looks pretty good, there are only a few details left to refine, but I wonder what instructions do one has to follow to get the Freesound key? The [OAuth2](https://freesound.org/docs/api/authentication.html#oauth2-authentication) steps? or maybe is the [Token](https://freesound.org/docs/api/authentication.html#token-authentication) enough?

File: openverse_catalog/dags/provider_api_scripts/freesound.py
Comment (krysal) [reply openverse_catalog/dags/provider_api_scripts/freesound.py]: My first thought is that we're not really interested in preview audios and we're keeping the preview for the main one just to have something that users can hear from Openverse directly. So I was thinking of the same approach as @sarayourfriend here. 

> Should the preview URL go in audio_url on the main object and then alt_files be reserved exclusively for downloadable formats...?

In my view, yes. This way we don't have to come up with an extra strategy to differentiate if files to download are previews or the full audio version.

**Comment by AetherUnbound** [comment]
> Should we have a more explicit error message when authentication fails? CC @AetherUnbound. Here's the current output:

Hmm, this is definitely an improvement we could add onto the `requester` object. Possibly that it would catch 401s and fail-fast? Did the script continue after that and try to run more requests or did it stop? I don't think there's a case where a request would 401 in one instance and 200 in another.

**Comment by zackkrida** [comment]
> > Should we have a more explicit error message when authentication fails? CC @AetherUnbound. Here's the current output:
> 
> Hmm, this is definitely an improvement we could add onto the `requester` object. Possibly that it would catch 401s and fail-fast? Did the script continue after that and try to run more requests or did it stop? I don't think there's a case where a request would 401 in one instance and 200 in another.

@AetherUnbound the script did stop after the failure.

File: openverse_catalog/dags/provider_api_scripts/freesound.py
Comment (zackkrida) [reply openverse_catalog/dags/provider_api_scripts/freesound.py]: The approach advocated for by @krysal and @sarayourfriend makes sense to me too; the script works great and I'll approve once this lands!

Comment (obulat) [reply openverse_catalog/dags/provider_api_scripts/freesound.py]: So, I will leave the hq mp3 file preview as the main file, and one alt_file (the original that can be downloaded by an authenticated user), right, @zackkrida?
We could also add an ogg file preview to the alt_files. I think it should be possible to download it  (even without authentication)

File: openverse_catalog/dags/providers/provider_api_scripts/freesound.py
Comment (obulat) [reply openverse_catalog/dags/providers/provider_api_scripts/freesound.py]: I've added the date :)

File: openverse_catalog/dags/providers/provider_api_scripts/freesound.py
Comment (obulat) [reply openverse_catalog/dags/providers/provider_api_scripts/freesound.py:186]: Line 164 replaces the main_audio object with the hq mp3 preview and its properties. Or, in an unlikely case previews don't exist, it doesn't.

File: openverse_catalog/dags/provider_api_scripts/freesound.py
Comment (zackkrida) [reply openverse_catalog/dags/provider_api_scripts/freesound.py]: @obulat is the ogg preview full length? my understanding is that we want `alt_files` to only be links to full-length, downloadable files.

Comment (sarayourfriend) [reply openverse_catalog/dags/provider_api_scripts/freesound.py]: Even if it is full-length, it seems like a betrayal of the upstream provider to circumvent their requirement for authentication before downloading that way (even though it's trivial to download the preview if you have certain technical competencies).

Comment (zackkrida) [reply openverse_catalog/dags/provider_api_scripts/freesound.py]: That's a good point @sarayourfriend 👍

Comment (obulat) [reply openverse_catalog/dags/provider_api_scripts/freesound.py]: I checked, and the preview files are full-length, but to use the API responsively, I will just leave the `main_audio` file in whichever format it is as the only `alt_file`, and have the `hq_mp3` format as the main file.

**Review by zackkrida — APPROVED** [review]
🥳

File: openverse_catalog/dags/providers/provider_api_scripts/freesound.py
Comment (krysal) [reply openverse_catalog/dags/providers/provider_api_scripts/freesound.py:187]: I didn't notice this a was a comment for the next line, with the unpacking operator, my bad, it's all good 👌

File: openverse_catalog/dags/providers/provider_api_scripts/freesound.py
Comment (krysal) [inline openverse_catalog/dags/providers/provider_api_scripts/freesound.py:33]: The key needs to be added to the env.template file.

**Review by krysal — APPROVED** [review]
LGTM! ✨

File: openverse_catalog/dags/providers/provider_api_scripts/freesound.py
Comment (obulat) [reply openverse_catalog/dags/providers/provider_api_scripts/freesound.py:33]: Great catch, @krysal , thank you!

**Pull request merged**
